### DISCUSSION TITLE
Coordinate precision - default read vs MeshIO read

### DISCUSSION THREAD

**Discussion by derek-kinakin** [pyvista/pyvista#5799] [Show and tell]
### Describe something you've made

I wanted to share some observations around reading meshes and coordinate precision issues (I believe). My hope is to learn more about the internals of PyVista and how to work through these issues in the future.

### Links to Documentation, Examples, or Class Definitions.

https://docs.pyvista.org/version/stable/api/utilities/_autosummary/pyvista.read.html#pyvista.read

### Pseudocode or Screenshots

Tested PyVista versions: 0.37 and 0.43.3
 
We often receive meshes created from drone photogrammetry data in variety of file formats and geographic coordinate systems. Once in while we get meshes that appear to have coordinate precision artifacts. For example here's a mesh in CloudCompare at its position in geographic space: 
![cc_not_translated](https://github.com/pyvista/pyvista/assets/89031238/0c9822f2-af99-4904-b6a4-0ca594ad57f2)

CloudCompare knows that these coordinate issues can occur and provides an option to translate/shift the mesh on loading, resulting in the correct looking product:
![cc_translated](https://github.com/pyvista/pyvista/assets/89031238/98e37fe3-581a-4d16-a913-43b109110b6d)

Opening the same mesh via PyVista's "read", with no other instructions, results in artifacts again. Translating/shifting the mesh after loading doesn't appear fix the issue. What does make a difference is having MeshIO available and calling "read" with the "file_format" option. This appears to force the reader to use MeshIO. Here's some simple code and a screenshot of the results:

```python
import pyvista as pv

mesh_path = 'test_mesh.ply'

default_read = pv.read(mesh_path)
mi_read = pv.read(mesh_path, file_format='ply')

pl = pv.Plotter(shape=(1,2), lighting='three lights')
pl.subplot(0,0)
pl.add_mesh(default_read, color='grey')
pl.add_text('Default', position='upper_left')

pl.subplot(0,1)
pl.add_mesh(mi_read, color='grey')
pl.add_text('MeshIO', position='upper_left')

pl.show()
```
![pv_v0_43](https://github.com/pyvista/pyvista/assets/89031238/1e4cdab5-39be-497b-8008-78a1d59c746f)

Other than eliminating the visual artifacts, the two read methods appear to return different types of objects. The default "read" returns a PolyData object. The MeshIO read returns an UnstructuredGrid:
![objects](https://github.com/pyvista/pyvista/assets/89031238/bde8f0c6-e1f3-4bac-bc6c-ee5ea15facc1)

In most of my code for these types of applications, I'm moving over to the MeshIO based read to avoid these type of coordinate precision issues. I'm keen to hear about other ways people tackle these types of problems.

Thanks for reading!

**Comment by darikg** [comment]
Looks like `vtkPLYReader` loads points as [floats](https://github.com/Kitware/VTK/blob/78825dc3982054481e4820c7d0df0c95ac4773a1/IO/PLY/vtkPLYReader.cxx#L103). Can you confirm that `default_read.points.dtype` is `float32` and  `mi_read.points.dtype` is `float64`? If that's the case the only option is a feature request to VTK to support variable precision like a lot of their filters. I do wonder about what CloudCompare is doing there though--that loss of precision is irreversible, so it must just be doing some kind of smoothing to make things look nice.

**Reply by derek-kinakin** [reply]
Looks like the types are what you expect:
![image](https://github.com/pyvista/pyvista/assets/89031238/374c406e-7cf1-47ec-982b-e73376c9498a)

I don't know what CloudCompare does in their backend. But, here's their description of the coordinate shift:
https://www.cloudcompare.org/doc/wiki/index.php/Global_Shift_and_Scale
